File: src/main/java/pcd/ass01/BoidsModel.java
package pcd.ass01;

import java.util.ArrayList;
import java.util.List;
import java.util.concurrent.locks.Lock;
import java.util.concurrent.locks.ReentrantLock;

public class BoidsModel {
    
    private List<Boid> boids;
    private double separationWeight; 
    private double alignmentWeight; 
    private double cohesionWeight; 
    private final double width;
    private final double height;
    private final double maxSpeed;
    private final double perceptionRadius;
    private final double avoidRadius;
    private volatile boolean isRunning;
    private volatile boolean isSuspended;
    private int nBoids;
    private final Lock separationWeightLock;
    private final Lock alignmentWeightLock;
    private final Lock cohesionWeightLock;

    public BoidsModel(int nBoids,
                            double initialSeparationWeight,
    						double initialAlignmentWeight,
    						double initialCohesionWeight,
    						double width,
    						double height,
    						double maxSpeed,
    						double perceptionRadius,
    						double avoidRadius){
        this.separationWeight = initialSeparationWeight;
        this.alignmentWeight = initialAlignmentWeight;
        this.cohesionWeight = initialCohesionWeight;
        this.width = width;
        this.height = height;
        this.maxSpeed = maxSpeed;
        this.perceptionRadius = perceptionRadius;
        this.avoidRadius = avoidRadius;
        this.boids = new ArrayList<>();
        this.nBoids = nBoids;
        this.isRunning = false;
        this.isSuspended = false;
        this.separationWeightLock = new ReentrantLock();
        this.alignmentWeightLock = new ReentrantLock();
        this.cohesionWeightLock = new ReentrantLock();
    }

    public boolean isRunning() {
        return isRunning;
    }

    public void turnOn() {
        isRunning = true;
    }

    public void turnOff() {
        isRunning = false;
    }

    public boolean isSuspended() {
        return isSuspended;
    }

    public void suspend() {
        this.isSuspended = true;
    }

    public void resume() {
        this.isSuspended = false;
    }

    public void generateBoids() {
        for (int i = 0; i < nBoids; i++) {
            P2d pos = new P2d(-width /2 + Math.random() * width, -height /2 + Math.random() * height);
            V2d vel = new V2d(Math.random() * maxSpeed /2 - maxSpeed /4, Math.random() * maxSpeed /2 - maxSpeed /4);
            this.boids.add(new Boid(pos, vel));
        }
    }

    public void clearBoids() {
        this.nBoids = 0;
        this.boids = new ArrayList<>();
    }
    
    public List<Boid> getBoids(){
    	return boids;
    }

    public double getMinX() {
    	return -width/2;
    }

    public double getMaxX() {
    	return width/2;
    }

    public double getMinY() {
    	return -height/2;
    }

    public double getMaxY() {
    	return height/2;
    }
    
    public double getWidth() {
    	return width;
    }
 
    public double getHeight() {
    	return height;
    }

    public void setSeparationWeight(double value) {
        this.separationWeightLock.lock();
        this.separationWeight = value;
        this.separationWeightLock.unlock();
    }

    public void setAlignmentWeight(double value) {
        this.alignmentWeightLock.lock();
        this.alignmentWeight = value;
        this.alignmentWeightLock.unlock();
    }

    public void setCohesionWeight(double value) {
        this.cohesionWeightLock.lock();
        this.cohesionWeight = value;
        this.cohesionWeightLock.unlock();
    }

    public double getSeparationWeight() {
    	return separationWeight;
    }

    public double getCohesionWeight() {
    	return cohesionWeight;
    }

    public double getAlignmentWeight() {
    	return alignmentWeight;
    }
    
    public double getMaxSpeed() {
    	return maxSpeed;
    }

    public double getAvoidRadius() {
    	return avoidRadius;
    }

    public double getPerceptionRadius() {
    	return perceptionRadius;
    }

    public void setBoidsNumber(int n) {
        this.nBoids = n;
    }
}

File: src/main/scala/pcd/ass01/BoidsModel.scala
package pcd.ass01

class BoidsModel(
                  var nBoids: Int,
                  var separationWeight: Double,
                  var alignmentWeight: Double,
                  var cohesionWeight: Double,
                  val width: Double, val height: Double,
                  val maxSpeed: Double,
                  val perceptionRadius: Double,
                  val avoidRadius: Double) {
  var boids: List[Boid] = List[Boid]()
  var isRunning = false
  var isSuspended = false

  def turnOn(): Unit = isRunning = true

  def turnOff(): Unit = isRunning = false

  def suspend(): Unit = isSuspended = true

  def resume(): Unit = isSuspended = false

  def generateBoids(): Unit =
    for (i <- 0 until nBoids) {
      val pos = new P2d(-width / 2 + Math.random * width, -height / 2 + Math.random * height)
      val vel = new V2d(Math.random * maxSpeed / 2 - maxSpeed / 4, Math.random * maxSpeed / 2 - maxSpeed / 4)
      boids = boids :+ Boid(pos, vel)
    }

  def clearBoids(): Unit =
    nBoids = 0
    boids = List[Boid]()

  def getMinX: Double = -width / 2
  def getMaxX: Double = width / 2
  def getMinY: Double = -height / 2
  def getMaxY: Double = height / 2

  def setSeparationWeight(value: Double): Unit = separationWeight = value
  def setAlignmentWeight(value: Double): Unit = alignmentWeight = value
  def setCohesionWeight(value: Double): Unit = cohesionWeight = value

  def setBoidsNumber(n: Int): Unit = nBoids = n
}